byzcoin/service.go

```diff
@@ -653,26 +677,56 @@ func isViewChangeTx(txs TxResults) *viewchange.View {
 	return req.GetView()
 }
 
-// GetCollectionView returns a read-only accessor to the collection
-// for the given skipchain.
-func (s *Service) GetCollectionView(scID skipchain.SkipBlockID) CollectionView {
-	cdb := s.getCollection(scID)
-	return &roCollection{cdb.coll}
+// GetReadOnlyStateTrie returns a read-only accessor to the trie for the given
+// skipchain.
+func (s *Service) GetReadOnlyStateTrie(scID skipchain.SkipBlockID) ReadOnlyStateTrie {
+	return s.getStateTrie(scID)
 }
 
-func (s *Service) getCollection(id skipchain.SkipBlockID) *collectionDB {
+func (s *Service) getStateTrie(id skipchain.SkipBlockID) *StateTrie {
+	if len(id) == 0 {
+		return nil
+	}
 	s.storage.Mutex.Lock()
 	defer s.storage.Mutex.Unlock()
 	idStr := fmt.Sprintf("%x", id)
-	col := s.collectionDB[idStr]
+	col := s.stateTries[idStr]
 	if col == nil {
 		db, name := s.GetAdditionalBucket([]byte(idStr))
-		s.collectionDB[idStr] = newCollectionDB(db, name)
-		return s.collectionDB[idStr]
+		_, nameIndex := s.GetAdditionalBucket([]byte(idStr + "_index"))
+		st, err := LoadStateTrie(db, name, nameIndex)
+		if err != nil {
+			log.Error(s.ServerIdentity(), idStr, err)
+			return nil
+		}
+		s.stateTries[idStr] = st
+		return s.stateTries[idStr]
 	}
 	return col
 }
 
+func (s *Service) createStateTrie(id skipchain.SkipBlockID, nonce []byte) *StateTrie {
+	if len(id) == 0 {
+		return nil
+	}
+	s.storage.Mutex.Lock()
+	defer s.storage.Mutex.Unlock()
+	idStr := fmt.Sprintf("%x", id)
+	if s.stateTries[idStr] != nil {
+		// Usually this function shouldn't be called if the trie
+		// already exists, so we return early.
+		return nil
+	}
+	db, name := s.GetAdditionalBucket([]byte(idStr))
+	st, err := NewStateTrie(db, name, nonce)
+	if err != nil {
+		log.Error(s.ServerIdentity(), idStr, err)
+		return nil
```

Comment:
in this case, the `nil` means an error, i.e., we failed to load/create the state trie. i kept it this way because the old `getCollections` was also doing it this way